qcodes/instrument_drivers/rigol/DS4034.py

```diff
@@ -0,0 +1,112 @@
+"""
+The QCodes driver of the oscilloscope Rigol DS4034. Please see
+
+http://int.rigol.com/File/TechDoc/20160831/DS4000E_ProgrammingGuide_EN.pdf
+
+for the programming manual for this device
+"""
+
+import numpy as np
+
+from qcodes import VisaInstrument
+from qcodes.utils.validators import Ints, Bool
+from qcodes.instrument.channel import InstrumentChannel, ChannelList
+
+
+class Rigol_DS4035_Channel(InstrumentChannel):
+
+    def __init__(self, parent, name, channel_number):
+        super().__init__(parent, name)
+
+        self.add_parameter(
+            "amplitude",
+            get_cmd=":meas:vamp? chan{}".format(channel_number)
+        )
+
+        self.add_parameter(
+            "vertical_scale",
+            get_cmd="chan{}:scale?".format(channel_number),
+            set_cmd="chan{}:scale ".format(channel_number) + "{}",
+            get_parser=float
+        )
+
+
+class Rigol_DS4035(VisaInstrument):
+    def __init__(self, name, address, **kwargs):
+        super().__init__(name, address, terminator='\n', **kwargs)
+
+        self.add_parameter(
+            "data_format",
+            get_cmd=":wav:form?",
+            set_cmd=":wav:form {}",
+            val_mapping={
+                'ascii': 'ascii',
+                'binary': 'binary'}
+        )
+
+        self.add_parameter(
+            "time_base",
+            get_cmd=":tim:scal?",
+            get_parser=float,
+            set_cmd=":tim:scal {}",
+            unit="s/sample"
+        )
+
+        self.add_parameter(
+            "sample_point_count",
+            get_cmd=":wav:poin?",
+            get_parser=int,
+            set_cmd=":wav:poin {}",
+            vals=Ints(min_value=0, max_value=1400)
+        )
+
+        self.add_parameter(
+            "source_channel",
+            set_cmd="WAV:SOUR CHAN{}",
+            vals=Ints(min_value=0, max_value=4)
+        )
+
+        self.add_parameter(
+            "mode",
+            set_cmd=":wav:mode {}",
+            val_mapping={
+                'normal': 'norm'}
+        )
+
+        self.add_parameter(
+            "enable_auto_scale",
+            set_cmd=":syst:aut {}",
+            vals=Bool(),
+            get_cmd=":syst:aut?"
+        )
+
+        self.add_parameter(
+            "auto_scale",
+            set_cmd=":AUToscale",
+            vals=Bool()
+        )
+
+        channels = ChannelList(self, "Channels", Rigol_DS4035_Channel, snapshotable=False)
+
+        for channel_number in range(1, 5):
+            channel = Rigol_DS4035_Channel(self, "ch{}".format(channel_number), channel_number)
+            channels.append(channel)
+
+        channels.lock()
+        self.add_submodule('channels', channels)
+
+    def get_wave_form(self, n_samples, channel):
```

Comment:
I still think that the waveform should be an ArrayParameter. Then it fits nicely into a `Loop`.